### ISSUE TITLE
non-blocking near-realtime indexing

### ISSUE THREAD

**Issue by darkskygit** [toeverything/OctoBase#217]
At present, the index behavior is designed to be executed when the search interface is called, which will block the search interface in a large workspace.

We should trigger the index or add the content to the index queue when the doc is modified. When the search interface is called, we should always search for the indexed data.

Previously, the internal data of yrs could not be accessed safely across threads, which was improved in new version. We should adjust the design of the search function to optimize performance.

**Comment by water-in-stone** [comment]
Does the issue persist as a bug? Furthermore, I aspire to contribute code to the development of OctoBase.

**Comment by darkskygit** [comment]
> Does the issue persist as a bug? Furthermore, I aspire to contribute code to the development of OctoBase.

Yes, the problem still exists! However we have recently rewritten the [implementation](https://github.com/toeverything/OctoBase/blob/master/libs/jwst-codec) to be compatible with yjs, it can be safely called by multiple threads, so we can transform the full-text index into a non-blocking implementation.

In addition, this issue is also related to another issue: https://github.com/toeverything/OctoBase/issues/508
